### DISCUSSION TITLE
8-bit versus 24-bit color prompt

### DISCUSSION THREAD

**Discussion by hastiin-manuelito** [xonsh/xonsh#4774] [General]
From what I can tell (by reading through the 'color_tools.py' and 'ansi_colors.py' modules) there are only 256 colors (8-bit) supported in the prompt, and any of the other 256 x 256 x 256 colors are shifted to one of these values. So, there is really no support for "#HEX colors of 6 characters in length" in the PROMPT as suggested in https://xon.sh/tutorial.html. However, in xterm, all 256 x 256 x 256 colors are supported in the prompt. This allows me to produce a nice (to me anyway) color gradient that I was trying achieve in xonsh, but failed to do so. 

I like using xonsh as I am learning Python and it works very well with matplotlib and other such scientific utilities, so it would be nice for this little bit flexibility. I know it doesn't break anything not to have this, but it would be nice either way.

So, is there a reason why a full 256 x 256 x 256 color palette is not supported in the PROMPT?

A snapshot of what I like (from xterm).

![2022-04-21-130621_925x151_scrot](https://user-images.githubusercontent.com/78440845/164524182-97a91c7f-38bc-4942-bb51-72c9d2dcbe1a.png)

Thanks.

**Comment by dyuri** [comment]
If you are sure that you terminal supports 24 bit colors, you can set `$PROMPT_TOOLKIT_COLOR_DEPTH` to `DEPTH_24_BIT`, and it won't use the nearest color from the 256 color palette, but the original one.

**Reply by dyuri** [reply]
I've found it: https://xon.sh/envvars.html#prompt-toolkit-color-depth

**Reply by hastiin-manuelito** [reply]
Tried this and no change to the color depth in the prompt. I use kitty term with 'truecolor' support enabled. This requires more investigation on my part; this is not by any means a serious issue for most people. 

Thanks either way.

**Reply by dyuri** [reply]
Well, this is strange indeed. `print_color()` handles the color codes properly, but they are wrong in the prompt:
![image](https://user-images.githubusercontent.com/126637/167085168-c856364b-626e-4d6d-b44b-733b1da840b5.png)

Is this what happening at your side too?

**Reply by hastiin-manuelito** [reply]
Yeah, that same thing happens to me where I end up with black bg's. I know that xonsh (using 'echo' and escape codes) does print all "truecolors". I didn't think to use `print_color()` using the same color substitution syntax as the prompt. I couldn't get anything to color with print_color, but now I know why.

Sometimes I would get color in the prompt, but they would be from one of the 256 8-bit colors. 

The logic in the files I originally referenced is set up to accept any truecolor, but it ultimately changes them into one of the 8-bit colors using a using a geometric minimum, I believe. 

As sort of an aside, I think ansi color support should entail using the 256 values that are "prepackaged" with any modern terminal emulators, for example, this set of escape codes: `\033[48;5;<N>` where N is in a range from 16-256 for backgrounds, similarly for foregrounds. Nowadays people just tend to override the first sixteen colors with any truecolor, using Xresources for example.

**Reply by dyuri** [reply]
You can still use truecolor ANSI escape codes without redefining the default ones in Xresources, that should work:
![image](https://user-images.githubusercontent.com/126637/167127727-3555990a-b28b-4026-ad83-54a85c91d68f.png)

But this is of course a bug anyway.

**Reply by hastiin-manuelito** [reply]
Definitely works in the xonsh output, but not in the prompt. I've no trouble with colors in xonsh outside the prompt.

**Reply by dyuri** [reply]
Well, I'm also using `kitty` and it works for me:
![image](https://user-images.githubusercontent.com/126637/167129271-49ca555d-9d00-4c03-88e7-28df29e2725b.png)
(Checked the colors using color picker.)

**Reply by hastiin-manuelito** [reply]
That `repacolor` tool seems to be a nice piece of software! I haven't used the `\x1b` escape sqeuence in the prompt, but I have used `\033`. Is there a difference? I know I entered the latter into the $PROMPT command and printed it to standard out, but it printed with `\x1b`.

edit: I just realized that you used, explicitly, the f character for formatting. I'm not at my computer now, but I can definitely check it out sooner than later.

**Reply by dyuri** [reply]
`\033` and `\x1b` should be the same, the escape code of the `escape` character (which is `27 = 0o33 = 0x1b`).

And thanks,`repacolor` is one of my favorite pet projects, with a long TODO list ;) (and as always, I'm considering a complete rewrite...).
